<a href="https://colab.research.google.com/github/AchmadAinunNaim/Machlearning_Semester5-/blob/main/Minggu4/Tugas_Jobsheet_04_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Tugas Jobsheet04**

1. Buatlah model klasifikasi dengan menggunakan SVM untuk data suara, voice.csv.

import data

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

load data


In [8]:
# Load Data
df = pd.read_csv('/content/voice.csv')
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

menampilkan Akurasi dan confusion matrikx

In [9]:
# Persiapan Data
X = df.drop('label', axis=1)
y = df['label']

# Pisahkan data menjadi data pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Buat dan Latih Model SVM
svm_model = SVC(kernel='linear')  # Gunakan kernel linear sebagai contoh
svm_model.fit(X_train, y_train)

# Evaluasi Model
y_pred = svm_model.predict(X_test)

# Hitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy}')

# Tampilkan laporan klasifikasi
print(classification_report(y_test, y_pred))

# Tampilkan matriks kebingungan
confusion = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(confusion)

Akurasi: 0.9416403785488959
              precision    recall  f1-score   support

      female       0.98      0.89      0.93       297
        male       0.91      0.99      0.95       337

    accuracy                           0.94       634
   macro avg       0.95      0.94      0.94       634
weighted avg       0.94      0.94      0.94       634

Confusion Matrix:
[[265  32]
 [  5 332]]


2. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
- Menggunakan data spam.csv
- Fitur CountVectorizer dengan mengaktifkan stop_words
- Evaluasi hasilnya

memasukkan data spam.csv

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Load Data
data = pd.read_csv('/content/spam.csv', encoding='latin-1')  # Pastikan menggunakan encoding yang sesuai
data.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

Drop 3 kolom dan cek data

In [19]:
# Drop 3 kolom terakhir dengan fungsi iloc
data = data.drop(data.iloc[:,2:], axis=1)

# Cek data
data.head()

v1                                                 v2
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

data untuk rename kolam v1 dan v2   dan cek data

In [20]:
# Data untuk rename kolom v1 dan v2
new_cols = {
    'v1': 'Labels',
    'v2': 'SMS'
}

# Rename nama kolom v1 dan v2
data = data.rename(columns=new_cols)

# cek data
data.head()


Labels                                                SMS
0    ham  Go until jurong point, crazy.. Available only ...
1    ham                      Ok lar... Joking wif u oni...
2   spam  Free entry in 2 a wkly comp to win FA Cup fina...
3    ham  U dun say so early hor... U c already then say...
4    ham  Nah I don't think he goes to usf, he lives aro...

akuransi

In [22]:
# Persiapan Data
X = data['SMS'].values
y = data['Labels'].values

# Pisahkan data menjadi data pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Buat dan konfigurasi CountVectorizer dengan stop_words
count_vectorizer = CountVectorizer(stop_words='english')

# Transformasi data teks menjadi vektor fitur
X_train_vectorized = count_vectorizer.fit_transform(X_train)
X_test_vectorized = count_vectorizer.transform(X_test)

# Buat dan Latih Model Multinomial Naive Bayes
nb_model = MultinomialNB()
nb_model.fit(X_train_vectorized, y_train)

# Evaluasi Model
y_pred = nb_model.predict(X_test_vectorized)

In [23]:
# Hitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi: {accuracy}')

# Tampilkan laporan klasifikasi
print(classification_report(y_test, y_pred))

# Tampilkan matriks kebingungan
confusion = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(confusion)

Akurasi: 0.9838565022421525
              precision    recall  f1-score   support

         ham       0.99      0.99      0.99       965
        spam       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115

Confusion Matrix:
[[959   6]
 [ 12 138]]


3. Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,
- Menggunakan data spam.csv
- Fitur TF-IDF dengan mengaktifkan stop_words
- Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.
- Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

In [25]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Persiapan Data
X = data['SMS']
y = data['Labels']

# Pisahkan data menjadi data pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Buat dan konfigurasi TF-IDF Vectorizer dengan stop_words
tfidf_vectorizer = TfidfVectorizer(stop_words='english')

# Transformasi data teks menjadi vektor fitur TF-IDF
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

# Buat dan Latih Model Multinomial Naive Bayes
nb_model = MultinomialNB()
nb_model.fit(X_train_tfidf, y_train)

# Evaluasi Model
y_pred = nb_model.predict(X_test_tfidf)

# Hitung akurasi
accuracy = accuracy_score(y_test, y_pred)
print(f'Akurasi (TF-IDF): {accuracy}')

# Tampilkan laporan klasifikasi
print(classification_report(y_test, y_pred))

# Tampilkan matriks kebingungan
confusion = confusion_matrix(y_test, y_pred)
print('Confusion Matrix (TF-IDF):')
print(confusion)

Akurasi (TF-IDF): 0.9668161434977578
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       965
        spam       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115

Confusion Matrix (TF-IDF):
[[965   0]
 [ 37 113]]


### **Kesimpulan**
Dalam kasus ini, baik model yang menggunakan CountVectorizer maupun model yang menggunakan TF-IDF Vectorizer memberikan tingkat akurasi yang sama, yaitu sekitar 96.68%. Oleh karena itu, dari segi tingkat akurasi, keduanya memiliki kinerja yang setara dalam memprediksi data "spam.csv".